crypto/kms/pubkeystore.go

```diff
@@ -135,54 +144,163 @@ func GetPublicKey(id proto.NodeID) (publicKey *ec.PublicKey, err error) {
+						C: nodeInfoTypes.Nonce.C,
+						D: nodeInfoTypes.Nonce.D,
+					},
+				}
+			}
+		}
+
 		return err // return from View func
 	})
+	if err != nil {
+		log.Errorf("get node info failed: %s", err)
+	}
 	return
 }
 
-// SetPublicKey verify nonce and set Public Key
-func SetPublicKey(id proto.NodeID, nonce mine.Uint256, publicKey *ec.PublicKey) (err error) {
-	keyHash := mine.HashBlock(publicKey.SerializeCompressed(), nonce)
-	idHash, err := hash.NewHashFromStr(string(id))
+// GetAllNodeID
+func GetAllNodeID() (nodeIDs []proto.NodeID, err error) {
+	err = (*bolt.DB)(pks.db).View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(pks.bucket)
+		if bucket == nil {
+			return ErrBucketNotInitialized
+		}
+		err := bucket.ForEach(func(k, v []byte) error {
+			nodeIDs = append(nodeIDs, proto.NodeID(k))
+			return nil
+		})
+
+		return err // return from View func
+	})
 	if err != nil {
-		return ErrNotValidNodeID
+		log.Errorf("get all node id failed: %s", err)
+	}
+	return
+
+}
+
+// SetPublicKey verifies nonce and set Public Key
+func SetPublicKey(id proto.NodeID, nonce mine.Uint256, publicKey *ec.PublicKey) (err error) {
+	nodeInfo := &proto.Node{
+		ID:        id,
+		Addr:      "",
+		PublicKey: publicKey,
+		Nonce:     nonce,
 	}
+	return SetNodeInfo(nodeInfo)
+}
 
-	if !keyHash.IsEqual(idHash) {
-		return ErrNodeIDKeyNonceNotMatch
+// SetNodeInfo verifies nonce and sets {proto.Node.ID: proto.Node}
+func SetNodeInfo(nodeInfo *proto.Node) (err error) {
+	if nodeInfo == nil {
+		return ErrNilNode
+	}
+	if !Unittest {
+		if nodeInfo.PublicKey == nil {
+			return ErrNilNode
+		}
+		keyHash := mine.HashBlock(nodeInfo.PublicKey.SerializeCompressed(), nodeInfo.Nonce)
+		id := nodeInfo.ID
+		idHash, err := hash.NewHashFromStr(string(id))
+		if err != nil {
+			return ErrNotValidNodeID
+		}
+		if !keyHash.IsEqual(idHash) {
+			return ErrNodeIDKeyNonceNotMatch
+		}
 	}
-	return setPublicKey(id, publicKey)
+
+	return setPublicKey(nodeInfo)
 }
 
 // setPublicKey sets id and its publicKey
-func setPublicKey(id proto.NodeID, publicKey *ec.PublicKey) (err error) {
-	return (*bolt.DB)(pks.db).Update(func(tx *bolt.Tx) error {
-		bucket := tx.Bucket(pks.bucket)
-		if bucket == nil {
-			return ErrBucketNotInitialized
-		}
-		return bucket.Put([]byte(id), publicKey.SerializeCompressed())
+func setPublicKey(nodeInfo *proto.Node) (err error) {
+	nodeBuf, err := pb.Marshal(&types.Node{
+		ID: &types.NodeID{
+			NodeID: string(nodeInfo.ID),
+		},
+		Addr: nodeInfo.Addr,
+		PublicKey: &types.PublicKey{
+			PublicKey: nodeInfo.PublicKey.SerializeCompressed(),
+		},
+		Nonce: &types.Nonce{
+			A: nodeInfo.Nonce.A,
+			B: nodeInfo.Nonce.B,
+			C: nodeInfo.Nonce.C,
+			D: nodeInfo.Nonce.D,
+		},
 	})
+	if err == nil {
+		return (*bolt.DB)(pks.db).Update(func(tx *bolt.Tx) error {
+			bucket := tx.Bucket(pks.bucket)
+			if bucket == nil {
+				return ErrBucketNotInitialized
+			}
+			return bucket.Put([]byte(nodeInfo.ID), nodeBuf)
+		})
+	} else {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
if block ends with a return statement, so drop this else and outdent its block